src/TesApi.Web/Scheduler.cs

```diff
@@ -204,36 +214,117 @@ private async Task ProcessTesTasksContinuouslyAsync(Expression<Func<TesTask, boo
+                    var completedTesTaskIds = new List<string>();
+
+                    while (await enumerator.MoveNextAsync())
+                    {
+                        // example: tasks-completions/12345678_12345678123456781234567812345678.json
+                        var blobName = enumerator.Current.Name;
+                        var json = (await container.GetBlobClient(blobName).DownloadContentAsync(stoppingToken)).Value.Content.ToString();
+                        var tesTaskCompletionMessage = JsonConvert.DeserializeObject<TesTaskCompletionMessage>(json);
+                        completedTesTaskIds.Add(tesTaskCompletionMessage.Id);
+                    }
+
+                    // Set class-level object so that it can be used by the existing task processor thread
+                    this.completedTesTaskIds = completedTesTaskIds;
+
+                    if (completedTesTaskIds.Count > 0)
+                    {
+                        logger.LogInformation($"{completedTesTaskIds.Count} TES task completions downloaded in {sw.Elapsed.TotalSeconds:n1}s; {sw.Elapsed.TotalSeconds / completedTesTaskIds.Count:n1} seconds per task completion");
+                    }
+
+                    await Task.Delay(runInterval, stoppingToken);                    
                 }
-                catch (RepositoryCollisionException exc)
+                catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
                 {
-                    logger.LogError(exc, $"RepositoryCollisionException in OrchestrateTesTasksOnBatch");
+                    logger.LogInformation($"ProcessTesTaskCompletionsContinuouslyAsync tasks cancelled.");
+                    break;
                 }
                 catch (Exception exc)
                 {
-                    logger.LogError(exc, "Updating TES Task '{TesTask}' threw {ExceptionType}: '{ExceptionMessage}'. Stack trace: {ExceptionStackTrace}", tesTask.Id, exc.GetType().FullName, exc.Message, exc.StackTrace);
+                    logger.LogError(exc, exc.Message);
+                    await Task.Delay(runInterval, stoppingToken);
                 }
+            }
+        }
 
-                if (!string.IsNullOrWhiteSpace(tesTask.PoolId) && (TesState.QUEUEDEnum == tesTask.State || TesState.RUNNINGEnum == tesTask.State))
+        private async Task ProcessTesTasks(List<TesTask> tesTasks, bool areExistingTesTasks, CancellationToken stoppingToken)
+        {
+            var pools = new HashSet<string>();
+            int completedTesTaskIdsCount = completedTesTaskIds.Count;
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                // Prioritize processing completed tasks, then terminal state tasks, then randomize order within each state to prevent head-of-line blocking
+                foreach (var tesTask in tesTasks.OrderByDescending(t => completedTesTaskIds.Contains(t.Id)).ThenBy(t => t.State).ThenBy(t => random.Next()))
```

Comment:
fyi; this is the main way completed TES tasks get prioritized and processed first